reframe/frontend/executors/__init__.py

```diff
@@ -197,6 +213,24 @@ def _environ_supported(self, check, environ):
         else:
             return ret and check.supports_environ(environ.name)
 
+    def _retry(self, checks, system):
+        while (self._stats.num_failures() and self._try_num <
+                                             self._max_retries):
+            self._try_num += 1
+            self._stats.next_try(self._try_num)
+            check_names_failed = self._stats.check_names_failed(try_num=-2)
```

Comment:
I would rewrite the whole loop as follows (more or less):

```python
while self._retry_count < self._num_retries:
    failed_checks = list(filter(lambda c: c if self._stats.has_failed(c.name)))
    if not failed_checks:
        break

    self._retry_count += 1
    self._stats.nextrun()
    self._printer.separator(...)
    self._runall(failed_checks, system)
```

Reply:
I adapted it. However, I don't think that replacing a proper while condition with a break statement inside the while loop makes it more clear...

Reply:
You noted "I would prefer sth more primitive. You get the failed tasks and the caller should do the filtering. After all, statistics care only about and manage the tasks". Therefore I created a function `self.tasks_failed` in `statistics` rather than `self.has_failed`, which would take a check name as argument and therefore the `statistics` would again have to deal with checks and not only tasks.

Reply:
Done.

Reply:
> I don't think that replacing a proper while condition with a break statement inside the while loop makes it more clear...

I disagree. There are cases that `break` statements gotos even make the code more clear and more efficient. The reason I used `break` is that I wanted to go over the tests once per loop iteration and not twice (count the failures first, then retrieve the failed tests). Instead, I retrieve once the failed tests and if none exists, I simply break.

Reply:
> The reason I used break is that I wanted to go over the tests once per loop iteration and not twice (count the failures first, then retrieve the failed tests). 
I see your point. However, if you want to do optimization, I would rather do it inside the `statistics`. There one could make sure that everything is computed in an optimal way. Yet, I doubt that the computation of the statistics is at all relevant for the performance of ReFrame. Running x tests should always take *much* more time than reading from x jobs whether it failed etc. So I would not trade in simplicity and clarity for unneeded performance optimizations.